test/utils.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * Turn a tree from the parser into a simple string format like PromQL(Expr(FunctionCall(...))) same as the way
+ * tests are written.
+ * @param tree
+ * @returns {string|undefined}
+ */
+export function treeToString(tree) {
+    let actual = ''
+    tree.iterate({
+        enter(type, start) {
+            if (!type.name) return
+            if (actual.length > 1 && actual.charAt(actual.length - 1) !== '(') {
+                // This is a case where we have sibling tokens so lets separate them,
+                actual += ', '
+            }
+            actual += type.name + '('
+            return undefined
+        },
+        leave(type, start) {
+            if (actual.charAt(actual.length - 1) === '(') {
+                // In case token is empty (no children) we don't want to render empty '()'
+                actual = actual.substring(0, actual.length - 1)
+            } else {
+                actual += ')'
+            }
+        }
+    })
+    return actual;
+}
+
+/**
+ * Returns a line context for the given file.
+ *
+ * @param {string} file
+ * @param {number} index
+ */
+function toLineContext(file, index) {
+    const endEol = file.indexOf('\n', index + 80);
+    const endIndex = endEol === -1 ? file.length : endEol;
+    return file.substring(index, endIndex).split(/\n/).map(str => '  | ' + str).join('\n');
+}
+
+export function fileTests(file, fileName) {
+    let caseExpr = /\s*#\s*(.*)(?:\r\n|\r|\n)([^]*?)==+>([^]*?)(?:$|(?:\r\n|\r|\n)+(?=#))/gy
+    let tests = []
+    let lastIndex = 0;
+    for (;;) {
+        let m = caseExpr.exec(file)
+        if (!m) throw new Error(`Unexpected file format in ${fileName} around\n\n${toLineContext(file, lastIndex)}`)
+
+        let [, name, configStr] = /(.*?)(\{.*?\})?$/.exec(m[1])
+        let config = configStr ? JSON.parse(configStr) : null
+
+        let text = m[2].trim(), expected = m[3].trim()
+        tests.push({
+            name,
+            run(parser) {
+                parser = parser.configure({strict: false, ...config})
+                const actual = treeToString(parser.parse(text))
```

Comment:
This is similar to the lezer utility function but instead of doing the test and compare during the tree walk, it now just converts the tree into string that can be compared directly with the expected string in the test.